core/src/main/java/com/linecorp/armeria/client/retry/RetryRuleBuilder.java

```diff
@@ -0,0 +1,285 @@
+     * has not been processed by the server. Therefore, you can safely retry the request without worrying about
+     * the idempotency of the request.
+     */
+    public RetryRuleBuilder onUnprocessed() {
+        return onException(UnprocessedRequestException.class);
+    }
+
+    /**
+     * Sets the {@linkplain Backoff#ofDefault() default backoff} and returns a newly created {@link RetryRule}.
+     */
+    public RetryRule thenBackoff() {
+        return thenBackoff(Backoff.ofDefault());
+    }
+
+    /**
+     * Sets the specified {@link Backoff} and returns a newly created {@link RetryRule}.
+     */
+    public RetryRule thenBackoff(Backoff backoff) {
+        requireNonNull(backoff, "backoff");
+        return build(RetryRuleDecision.retry(backoff));
+    }
+
+    /**
+     * Sets a {@link Backoff} which limits the number of attempts up to the specified value and
+     * never waits between attempts. Returns a newly created {@link RetryRule}.
+     */
+    public RetryRule thenImmediately(int maxAttempts) {
+        checkArgument(maxAttempts > 0, "maxAttempts: %s (expected: > 0)", maxAttempts);
+        final Backoff backOff = Backoff.withoutDelay().withMaxAttempts(maxAttempts);
+        return build(RetryRuleDecision.retry(backOff));
+    }
+
+    /**
+     * Returns a newly created {@link RetryRule} that never retries.
+     */
+    public RetryRule thenNoRetry() {
+        return build(RetryRuleDecision.noRetry());
+    }
+
+    private RetryRule build(RetryRuleDecision decision) {
+        final Set<HttpMethod> methods = Sets.immutableEnumSet(this.methods);
+        final Set<HttpStatusClass> statusClasses = Sets.immutableEnumSet(statusClassesBuilder.build());
+        final Set<HttpStatus> statuses = statusesBuilder.build();
+        final Predicate<HttpStatus> statusFilter = this.statusFilter;
+        final Predicate<Throwable> exceptionFilter = this.exceptionFilter;
+
+        if (decision != RetryRuleDecision.noRetry() && exceptionFilter == null && statusFilter == null &&
+            statuses.isEmpty() && statusClasses.isEmpty()) {
+            throw new IllegalStateException(
+                    "Should set at least one of status, status class and an expected exception type " +
+                    "if a backoff was set.");
+        }
+
+        return (ctx, cause) -> {
+            if (!methods.contains(ctx.request().method())) {
+                return NEXT;
+            }
+
+            if (cause != null && exceptionFilter != null && exceptionFilter.test(Exceptions.peel(cause))) {
+                return CompletableFuture.completedFuture(decision);
+            }
+
+            if (ctx.log().isAvailable(RequestLogProperty.RESPONSE_HEADERS)) {
+                final HttpStatus responseStatus = ctx.log().partial().responseHeaders().status();
+                if (statusClasses != null && statusClasses.contains(responseStatus.codeClass())) {
+                    return CompletableFuture.completedFuture(decision);
```

Comment:
Think you can move this out of the lambda instead of creating every time, similar to how `NEXT` is cached. Could also consider having a completed future created in the `RetryRuleDecision` constructor itself.

Reply:
> Think you can move this out of the lambda instead of creating every time

Good idea.

> Could also consider having a completed future created in the RetryRuleDecision constructor itself.

I couldn't follow this review for custom RetryRule. A user might want to lazily complete the decision to decide a rule.

Reply:
On seconds thought, I can add singleton `CompletableFuture<RetryRuleDecision>` for the default backoff.